newsrevenuehub/rev-engine
# Dev 1056 deliver initial confirmation emails for recurring donations mvp

Please fill out each section even if it's just with "N/A"

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

This is complete in terms of being ready for review, but do need to upload the new email template (/email_templates/nrh-default-contribution-confirmation-email.html) to Mailgun and confirm everything works as expected.

#### What's this PR do?

Causes contribution confirmation emails to be sent to contributors making recurring donations (vs. status quo where only contributors making one time donations get a confirmation email).

Along the way, a number of notable changes were made (some driven by the need to boost test coverage, which required touching otherwise untouched code).

- Retires apps.emails.models.PageTemplateEmail and related admin, signals, and management command files
- Enables eager celery task execution in tests
- Adds `ESP_TEMPLATE_ID_FOR_CONTRIBUTION_CONFIRMATION` to settings/base.py. This value is pulled from environment, so will need to be set up in Heroku when deploy. We'll get this value when we upload the template to Mailgun.
- Renames `send_donor_email` task to `send_email` task, since nothing about implementation was tied to contributors.
- Adds a new task `send_contribution_confirmation_email` which is used to trigger contribution confirmation emails. This allows us in short term to use a single email template to handle both recurring and one time contributions.
- Refactors the view tests around one-time and recurring payments to assert that send_contribution_confirmation_email gets called.
- Refactors validation around whether a user is permitted to reference a given foreign keyed object when creating or updating an instance of a resource.
  - This validator is `apps.common.validators.ValidateFkReferenceOwnership`.
  - This validator allowed `apps.api.permissions.HasCreatePrivilegesRole` and the `user_has_create_permission_by_virtue_of_role` method on some models to be retired.
- Adds `drf_extra_fields` as a dependency so we can use `PresentablePrimaryKeyRelatedField`, which provides a standardized way to refer to a model by primary key in update/create, while serializing in read operations (which was being handled inconsistently across different resources before this). This allowed retiring `apps.common.fileds.SerializedOnReadElsePk`
- Adds an .eslintrc to turn off some overly aggressive rules from testing-library which were errantly flagging the custom `cy.getByTestId` command as problematic.
- Updates some fixtures in line with recent api changes
- **Pins django-simple-history to 3.0.0**: This became necessary late in the game. There were strange problems happening in CI when `python manage.py makemigrations --dry-run --check` ran. Locally, that would not show any missing migration, but in CI it did. I discovered that reverting django-simple-history stopped that problem (without understanding underlying problem).  One failed run can be seen here: https://github.com/newsrevenuehub/rev-engine/runs/6560719461?check_suite_focus=true#step:9:50


#### Why are we doing this? How does it help us?

Business reasons are discussed in parent ticket.

More specifically, regarding retiring the PageTemplateEmail model, there was no clear use case for the abstraction provided by the model, and it was built in such a way that using a single template to handle both one-time and recurring contributions would go against the semantics of the code.  There's a fair amount of overhead to having this data model without any clear business value.

The advantage of the new approach is that it simplifies the connection with the ESP. The `send_contribution_confirmation_email` (as well as `send_email` which it calls internally) take a template identifier, a to email address, a subject, and template variables (a dict) that gets sent to the ESP, and the task gets called (with `.delay`) in the stripe payment function. 

#### How should this be manually tested?

- After uploading the template to MailGun, in the environment running the app, set an environment variable with the name `ESP_TEMPLATE_ID_FOR_CONTRIBUTION_CONFIRMATION` to the value of the uploaded MailGun email template identifier.
- Start the server (or restart after ensuring updated env is sourced)
- Visit a donation page and make a recurring contribution using an email address that you have access to.
- Check your inbox and confirm that a confirmation email came.


#### Have automated unit tests been added?

Not added, but updated to ensure task for sending confirmation emails gets called in both one time and recurring cases.

#### How should this change be communicated to end users?


#### Are there any smells or added technical debt to note?

There were two old migrations that were importing directly from apps.emails.models in order to dynamically set default values. In general, it's best to avoid importing models in this way into migrations because it can make it tricky to remove models. 

After removing the PageTemplateEmail and its models file, I was unable to make or run migrations because refs to the PageTemplateEmail model were broken. To get around this, I removed these references from the old migrations. This doesn't seem to create any problems locally, but changing old migrations is possibly tricky (in this case, I think it's fine).

Additionally, it became necessary to pin django-simple-history to 3.0.0 (see note above). At some point it may be worthwhile to run down that issue.


#### Has this been documented? If so, where?

#### What are the relevant tickets?
**(Note: Please use syntax [JIRA-123] to link any relevant tickets)**

[DEV-1056]

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

- Upload the new email template to MailGun and grab the template identifier
- Set `ESP_TEMPLATE_ID_FOR_CONTRIBUTION_CONFIRMATION` env var on Heroku to the value of the identifier

#### Are there next steps? If so, what? Have tickets been opened for them?
